File: app/airports.py
"""Airport resolution: accept a city/airport name or an IATA code, return IATA.

The frontend already submits clean 3-letter codes, but the API and CLI may
receive human input like "Dhaka" or "Hazrat Shahjalal". This module normalizes
either form to the IATA code that SerpAPI's ``departure_id``/``arrival_id`` need.

Any valid-looking 3-letter code is accepted as-is (SerpAPI supports codes beyond
this curated list); non-code input is matched against the table below.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

_CODE_RE = re.compile(r"^[A-Za-z]{3}$")


@dataclass(frozen=True)
class Airport:
    code: str
    name: str
    city: str
    country: str


# Kept in sync with frontend/src/lib/airports.ts.
AIRPORTS: list[Airport] = [
    Airport("DAC", "Hazrat Shahjalal International", "Dhaka", "Bangladesh"),
    Airport("CGP", "Shah Amanat International", "Chattogram", "Bangladesh"),
    Airport("JFK", "John F. Kennedy International", "New York", "United States"),
    Airport("EWR", "Newark Liberty International", "Newark", "United States"),
    Airport("LGA", "LaGuardia", "New York", "United States"),
    Airport("LAX", "Los Angeles International", "Los Angeles", "United States"),
    Airport("SFO", "San Francisco International", "San Francisco", "United States"),
    Airport("ORD", "O'Hare International", "Chicago", "United States"),
    Airport("ATL", "Hartsfield-Jackson", "Atlanta", "United States"),
    Airport("DFW", "Dallas/Fort Worth International", "Dallas", "United States"),
    Airport("SEA", "Seattle-Tacoma International", "Seattle", "United States"),
    Airport("MIA", "Miami International", "Miami", "United States"),
    Airport("BOS", "Logan International", "Boston", "United States"),
    Airport("IAD", "Washington Dulles International", "Washington", "United States"),
    Airport("DEN", "Denver International", "Denver", "United States"),
    Airport("LAS", "Harry Reid International", "Las Vegas", "United States"),
    Airport("YYZ", "Toronto Pearson International", "Toronto", "Canada"),
    Airport("YVR", "Vancouver International", "Vancouver", "Canada"),
    Airport("LHR", "Heathrow", "London", "United Kingdom"),
    Airport("LGW", "Gatwick", "London", "United Kingdom"),
    Airport("MAN", "Manchester", "Manchester", "United Kingdom"),
    Airport("CDG", "Charles de Gaulle", "Paris", "France"),
    Airport("ORY", "Orly", "Paris", "France"),
    Airport("AMS", "Schiphol", "Amsterdam", "Netherlands"),
    Airport("FRA", "Frankfurt International", "Frankfurt", "Germany"),
    Airport("MUC", "Munich International", "Munich", "Germany"),
    Airport("MAD", "Adolfo Suarez Madrid-Barajas", "Madrid", "Spain"),
    Airport("BCN", "Barcelona-El Prat", "Barcelona", "Spain"),
    Airport("FCO", "Leonardo da Vinci-Fiumicino", "Rome", "Italy"),
    Airport("MXP", "Milan Malpensa", "Milan", "Italy"),
    Airport("ZRH", "Zurich", "Zurich", "Switzerland"),
    Airport("VIE", "Vienna International", "Vienna", "Austria"),
    Airport("CPH", "Copenhagen", "Copenhagen", "Denmark"),
    Airport("ARN", "Stockholm Arlanda", "Stockholm", "Sweden"),
    Airport("OSL", "Oslo Gardermoen", "Oslo", "Norway"),
    Airport("HEL", "Helsinki-Vantaa", "Helsinki", "Finland"),
    Airport("DUB", "Dublin", "Dublin", "Ireland"),
    Airport("LIS", "Humberto Delgado", "Lisbon", "Portugal"),
    Airport("IST", "Istanbul", "Istanbul", "Turkey"),
    Airport("SVO", "Sheremetyevo", "Moscow", "Russia"),
    Airport("DXB", "Dubai International", "Dubai", "United Arab Emirates"),
    Airport("AUH", "Abu Dhabi International", "Abu Dhabi", "United Arab Emirates"),
    Airport("DOH", "Hamad International", "Doha", "Qatar"),
    Airport("RUH", "King Khalid International", "Riyadh", "Saudi Arabia"),
    Airport("JED", "King Abdulaziz International", "Jeddah", "Saudi Arabia"),
    Airport("KWI", "Kuwait International", "Kuwait City", "Kuwait"),
    Airport("DEL", "Indira Gandhi International", "Delhi", "India"),
    Airport("BOM", "Chhatrapati Shivaji Maharaj", "Mumbai", "India"),
    Airport("BLR", "Kempegowda International", "Bengaluru", "India"),
    Airport("MAA", "Chennai International", "Chennai", "India"),
    Airport("CCU", "Netaji Subhas Chandra Bose", "Kolkata", "India"),
    Airport("HYD", "Rajiv Gandhi International", "Hyderabad", "India"),
    Airport("KTM", "Tribhuvan International", "Kathmandu", "Nepal"),
    Airport("CMB", "Bandaranaike International", "Colombo", "Sri Lanka"),
    Airport("KHI", "Jinnah International", "Karachi", "Pakistan"),
    Airport("LHE", "Allama Iqbal International", "Lahore", "Pakistan"),
    Airport("ISB", "Islamabad International", "Islamabad", "Pakistan"),
    Airport("BKK", "Suvarnabhumi", "Bangkok", "Thailand"),
    Airport("DMK", "Don Mueang International", "Bangkok", "Thailand"),
    Airport("SIN", "Changi", "Singapore", "Singapore"),
    Airport("KUL", "Kuala Lumpur International", "Kuala Lumpur", "Malaysia"),
    Airport("CGK", "Soekarno-Hatta International", "Jakarta", "Indonesia"),
    Airport("DPS", "Ngurah Rai International", "Bali", "Indonesia"),
    Airport("MNL", "Ninoy Aquino International", "Manila", "Philippines"),
    Airport("HAN", "Noi Bai International", "Hanoi", "Vietnam"),
    Airport("SGN", "Tan Son Nhat International", "Ho Chi Minh City", "Vietnam"),
    Airport("HKG", "Hong Kong International", "Hong Kong", "Hong Kong"),
    Airport("TPE", "Taoyuan International", "Taipei", "Taiwan"),
    Airport("ICN", "Incheon International", "Seoul", "South Korea"),
    Airport("NRT", "Narita International", "Tokyo", "Japan"),
    Airport("HND", "Haneda", "Tokyo", "Japan"),
    Airport("KIX", "Kansai International", "Osaka", "Japan"),
    Airport("PEK", "Beijing Capital International", "Beijing", "China"),
    Airport("PVG", "Shanghai Pudong International", "Shanghai", "China"),
    Airport("CAN", "Guangzhou Baiyun International", "Guangzhou", "China"),
    Airport("SYD", "Kingsford Smith", "Sydney", "Australia"),
    Airport("MEL", "Melbourne", "Melbourne", "Australia"),
    Airport("BNE", "Brisbane", "Brisbane", "Australia"),
    Airport("AKL", "Auckland", "Auckland", "New Zealand"),
    Airport("JNB", "O. R. Tambo International", "Johannesburg", "South Africa"),
    Airport("CPT", "Cape Town International", "Cape Town", "South Africa"),
    Airport("CAI", "Cairo International", "Cairo", "Egypt"),
    Airport("NBO", "Jomo Kenyatta International", "Nairobi", "Kenya"),
    Airport("ADD", "Bole International", "Addis Ababa", "Ethiopia"),
    Airport("LOS", "Murtala Muhammed International", "Lagos", "Nigeria"),
    Airport("GRU", "Sao Paulo/Guarulhos", "Sao Paulo", "Brazil"),
    Airport("GIG", "Rio de Janeiro/Galeao", "Rio de Janeiro", "Brazil"),
    Airport("EZE", "Ministro Pistarini", "Buenos Aires", "Argentina"),
    Airport("SCL", "Arturo Merino Benitez", "Santiago", "Chile"),
    Airport("BOG", "El Dorado International", "Bogota", "Colombia"),
    Airport("LIM", "Jorge Chavez International", "Lima", "Peru"),
    Airport("MEX", "Benito Juarez International", "Mexico City", "Mexico"),
    Airport("CUN", "Cancun International", "Cancun", "Mexico"),
]

_BY_CODE = {a.code: a for a in AIRPORTS}
# First-listed airport wins for multi-airport cities (e.g. New York -> JFK).
_BY_CITY: dict[str, Airport] = {}
for _a in AIRPORTS:
    _BY_CITY.setdefault(_a.city.lower(), _a)


def resolve_airport(value: str) -> str:
    """Return the IATA code for ``value`` (a code, city, or airport name).

    Raises ``ValueError`` if the input cannot be mapped to an airport.
    """

    if value is None:
        raise ValueError("Airport is required")
    text = value.strip()
    if not text:
        raise ValueError("Airport is required")

    # A bare 3-letter code is accepted directly (SerpAPI supports any valid IATA).
    if _CODE_RE.match(text):
        return text.upper()

    q = text.lower()
    # Exact city match first, then substring against city / name / country.
    if q in _BY_CITY:
        return _BY_CITY[q].code
    for a in AIRPORTS:
        if q == a.name.lower():
            return a.code
    for a in AIRPORTS:
        if q in a.city.lower() or q in a.name.lower() or q in a.country.lower():
            return a.code

    raise ValueError(
        f"Could not resolve '{value}' to an airport. Use an IATA code (e.g. DAC) "
        "or a known city/airport name."
    )

File: frontend/src/lib/airports.ts
// Curated list of major world airports for the From/To autocomplete.
// Users can search by city, country, airport name, or IATA code; the form
// still submits the 3-letter IATA `code` that the backend / SerpAPI expects.

export interface Airport {
  code: string; // IATA (e.g. "DAC")
  name: string; // Airport name (e.g. "Hazrat Shahjalal International")
  city: string;
  country: string;
}

export const AIRPORTS: Airport[] = [
  { code: "DAC", name: "Hazrat Shahjalal International", city: "Dhaka", country: "Bangladesh" },
  { code: "CGP", name: "Shah Amanat International", city: "Chattogram", country: "Bangladesh" },
  { code: "JFK", name: "John F. Kennedy International", city: "New York", country: "United States" },
  { code: "EWR", name: "Newark Liberty International", city: "Newark", country: "United States" },
  { code: "LGA", name: "LaGuardia", city: "New York", country: "United States" },
  { code: "LAX", name: "Los Angeles International", city: "Los Angeles", country: "United States" },
  { code: "SFO", name: "San Francisco International", city: "San Francisco", country: "United States" },
  { code: "ORD", name: "O'Hare International", city: "Chicago", country: "United States" },
  { code: "ATL", name: "Hartsfield-Jackson", city: "Atlanta", country: "United States" },
  { code: "DFW", name: "Dallas/Fort Worth International", city: "Dallas", country: "United States" },
  { code: "SEA", name: "Seattle-Tacoma International", city: "Seattle", country: "United States" },
  { code: "MIA", name: "Miami International", city: "Miami", country: "United States" },
  { code: "BOS", name: "Logan International", city: "Boston", country: "United States" },
  { code: "IAD", name: "Washington Dulles International", city: "Washington", country: "United States" },
  { code: "DEN", name: "Denver International", city: "Denver", country: "United States" },
  { code: "LAS", name: "Harry Reid International", city: "Las Vegas", country: "United States" },
  { code: "YYZ", name: "Toronto Pearson International", city: "Toronto", country: "Canada" },
  { code: "YVR", name: "Vancouver International", city: "Vancouver", country: "Canada" },
  { code: "LHR", name: "Heathrow", city: "London", country: "United Kingdom" },
  { code: "LGW", name: "Gatwick", city: "London", country: "United Kingdom" },
  { code: "MAN", name: "Manchester", city: "Manchester", country: "United Kingdom" },
  { code: "CDG", name: "Charles de Gaulle", city: "Paris", country: "France" },
  { code: "ORY", name: "Orly", city: "Paris", country: "France" },
  { code: "AMS", name: "Schiphol", city: "Amsterdam", country: "Netherlands" },
  { code: "FRA", name: "Frankfurt International", city: "Frankfurt", country: "Germany" },
  { code: "MUC", name: "Munich International", city: "Munich", country: "Germany" },
  { code: "MAD", name: "Adolfo Suarez Madrid-Barajas", city: "Madrid", country: "Spain" },
  { code: "BCN", name: "Barcelona-El Prat", city: "Barcelona", country: "Spain" },
  { code: "FCO", name: "Leonardo da Vinci-Fiumicino", city: "Rome", country: "Italy" },
  { code: "MXP", name: "Milan Malpensa", city: "Milan", country: "Italy" },
  { code: "ZRH", name: "Zurich", city: "Zurich", country: "Switzerland" },
  { code: "VIE", name: "Vienna International", city: "Vienna", country: "Austria" },
  { code: "CPH", name: "Copenhagen", city: "Copenhagen", country: "Denmark" },
  { code: "ARN", name: "Stockholm Arlanda", city: "Stockholm", country: "Sweden" },
  { code: "OSL", name: "Oslo Gardermoen", city: "Oslo", country: "Norway" },
  { code: "HEL", name: "Helsinki-Vantaa", city: "Helsinki", country: "Finland" },
  { code: "DUB", name: "Dublin", city: "Dublin", country: "Ireland" },
  { code: "LIS", name: "Humberto Delgado", city: "Lisbon", country: "Portugal" },
  { code: "IST", name: "Istanbul", city: "Istanbul", country: "Turkey" },
  { code: "SVO", name: "Sheremetyevo", city: "Moscow", country: "Russia" },
  { code: "DXB", name: "Dubai International", city: "Dubai", country: "United Arab Emirates" },
  { code: "AUH", name: "Abu Dhabi International", city: "Abu Dhabi", country: "United Arab Emirates" },
  { code: "DOH", name: "Hamad International", city: "Doha", country: "Qatar" },
  { code: "RUH", name: "King Khalid International", city: "Riyadh", country: "Saudi Arabia" },
  { code: "JED", name: "King Abdulaziz International", city: "Jeddah", country: "Saudi Arabia" },
  { code: "KWI", name: "Kuwait International", city: "Kuwait City", country: "Kuwait" },
  { code: "DEL", name: "Indira Gandhi International", city: "Delhi", country: "India" },
  { code: "BOM", name: "Chhatrapati Shivaji Maharaj", city: "Mumbai", country: "India" },
  { code: "BLR", name: "Kempegowda International", city: "Bengaluru", country: "India" },
  { code: "MAA", name: "Chennai International", city: "Chennai", country: "India" },
  { code: "CCU", name: "Netaji Subhas Chandra Bose", city: "Kolkata", country: "India" },
  { code: "HYD", name: "Rajiv Gandhi International", city: "Hyderabad", country: "India" },
  { code: "KTM", name: "Tribhuvan International", city: "Kathmandu", country: "Nepal" },
  { code: "CMB", name: "Bandaranaike International", city: "Colombo", country: "Sri Lanka" },
  { code: "KHI", name: "Jinnah International", city: "Karachi", country: "Pakistan" },
  { code: "LHE", name: "Allama Iqbal International", city: "Lahore", country: "Pakistan" },
  { code: "ISB", name: "Islamabad International", city: "Islamabad", country: "Pakistan" },
  { code: "BKK", name: "Suvarnabhumi", city: "Bangkok", country: "Thailand" },
  { code: "DMK", name: "Don Mueang International", city: "Bangkok", country: "Thailand" },
  { code: "SIN", name: "Changi", city: "Singapore", country: "Singapore" },
  { code: "KUL", name: "Kuala Lumpur International", city: "Kuala Lumpur", country: "Malaysia" },
  { code: "CGK", name: "Soekarno-Hatta International", city: "Jakarta", country: "Indonesia" },
  { code: "DPS", name: "Ngurah Rai International", city: "Bali", country: "Indonesia" },
  { code: "MNL", name: "Ninoy Aquino International", city: "Manila", country: "Philippines" },
  { code: "HAN", name: "Noi Bai International", city: "Hanoi", country: "Vietnam" },
  { code: "SGN", name: "Tan Son Nhat International", city: "Ho Chi Minh City", country: "Vietnam" },
  { code: "HKG", name: "Hong Kong International", city: "Hong Kong", country: "Hong Kong" },
  { code: "TPE", name: "Taoyuan International", city: "Taipei", country: "Taiwan" },
  { code: "ICN", name: "Incheon International", city: "Seoul", country: "South Korea" },
  { code: "NRT", name: "Narita International", city: "Tokyo", country: "Japan" },
  { code: "HND", name: "Haneda", city: "Tokyo", country: "Japan" },
  { code: "KIX", name: "Kansai International", city: "Osaka", country: "Japan" },
  { code: "PEK", name: "Beijing Capital International", city: "Beijing", country: "China" },
  { code: "PVG", name: "Shanghai Pudong International", city: "Shanghai", country: "China" },
  { code: "CAN", name: "Guangzhou Baiyun International", city: "Guangzhou", country: "China" },
  { code: "SYD", name: "Kingsford Smith", city: "Sydney", country: "Australia" },
  { code: "MEL", name: "Melbourne", city: "Melbourne", country: "Australia" },
  { code: "BNE", name: "Brisbane", city: "Brisbane", country: "Australia" },
  { code: "AKL", name: "Auckland", city: "Auckland", country: "New Zealand" },
  { code: "JNB", name: "O. R. Tambo International", city: "Johannesburg", country: "South Africa" },
  { code: "CPT", name: "Cape Town International", city: "Cape Town", country: "South Africa" },
  { code: "CAI", name: "Cairo International", city: "Cairo", country: "Egypt" },
  { code: "NBO", name: "Jomo Kenyatta International", city: "Nairobi", country: "Kenya" },
  { code: "ADD", name: "Bole International", city: "Addis Ababa", country: "Ethiopia" },
  { code: "LOS", name: "Murtala Muhammed International", city: "Lagos", country: "Nigeria" },
  { code: "GRU", name: "Sao Paulo/Guarulhos", city: "Sao Paulo", country: "Brazil" },
  { code: "GIG", name: "Rio de Janeiro/Galeao", city: "Rio de Janeiro", country: "Brazil" },
  { code: "EZE", name: "Ministro Pistarini", city: "Buenos Aires", country: "Argentina" },
  { code: "SCL", name: "Arturo Merino Benitez", city: "Santiago", country: "Chile" },
  { code: "BOG", name: "El Dorado International", city: "Bogota", country: "Colombia" },
  { code: "LIM", name: "Jorge Chavez International", city: "Lima", country: "Peru" },
  { code: "MEX", name: "Benito Juarez International", city: "Mexico City", country: "Mexico" },
  { code: "CUN", name: "Cancun International", city: "Cancun", country: "Mexico" },
];

/** A short human-friendly label, e.g. "Dhaka, Bangladesh — Hazrat Shahjalal (DAC)". */
export function airportLabel(a: Airport): string {
  return `${a.city}, ${a.country} — ${a.name} (${a.code})`;
}

const AIRPORT_BY_CODE = new Map(AIRPORTS.map((a) => [a.code, a]));

export function findAirport(code: string): Airport | undefined {
  return AIRPORT_BY_CODE.get(code.trim().toUpperCase());
}

/** Rank airports against a free-text query. Returns up to `limit` matches. */
export function searchAirports(query: string, limit = 8): Airport[] {
  const q = query.trim().toLowerCase();
  if (!q) return AIRPORTS.slice(0, limit);

  const scored: { airport: Airport; score: number }[] = [];
  for (const a of AIRPORTS) {
    const code = a.code.toLowerCase();
    const city = a.city.toLowerCase();
    const country = a.country.toLowerCase();
    const name = a.name.toLowerCase();

    let score = 0;
    if (code === q) score = 100;
    else if (city === q) score = 90;
    else if (code.startsWith(q)) score = 80;
    else if (city.startsWith(q)) score = 70;
    else if (name.startsWith(q)) score = 55;
    else if (country.startsWith(q)) score = 45;
    else if (city.includes(q)) score = 40;
    else if (name.includes(q)) score = 30;
    else if (country.includes(q)) score = 20;
    else if (code.includes(q)) score = 15;

    if (score > 0) scored.push({ airport: a, score });
  }

  scored.sort((x, y) => y.score - x.score || x.airport.city.localeCompare(y.airport.city));
  return scored.slice(0, limit).map((s) => s.airport);
}
